clients/pkg/promtail/targets/file/decompresser.go

```diff
@@ -0,0 +1,285 @@
+		decompressLib = "bzip2"
+		reader = bzip2.NewReader(f)
+	}
+
+	level.Info(logger).Log("msg", fmt.Sprintf("using %q to decompress file %q", decompressLib, f.Name()))
+
+	if reader != nil {
+		return reader, nil
+	}
+
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	return nil, fmt.Errorf("file %q with unsupported extension", f.Name())
+}
+
+func (t *decompresser) updatePosition() {
+	positionSyncPeriod := t.positions.SyncPeriod()
+	positionWait := time.NewTicker(positionSyncPeriod)
+	defer func() {
+		positionWait.Stop()
+		level.Info(t.logger).Log("msg", "position timer: exited", "path", t.path)
+		close(t.posdone)
+	}()
+
+	for {
+		select {
+		case <-positionWait.C:
+			if err := t.MarkPositionAndSize(); err != nil {
+				level.Error(t.logger).Log("msg", "position timer: error getting position and/or size, stopping decompresser", "path", t.path, "error", err)
+				return
+			}
+		case <-t.posquit:
+			return
+		}
+	}
+}
+
+// readLines read all existing lines of the given compressed file.
+//
+// It first decompress the file as a whole using a reader and then it will iterate
+// over its chunks, separated by '\n'.
+// During each iteration, the parsed and decoded log line is then sent to the API with the current timestamp.
+func (t *decompresser) readLines() {
+	level.Info(t.logger).Log("msg", "read lines routine: started", "path", t.path)
+	t.running.Store(true)
+
+	defer func() {
+		t.cleanupMetrics()
+		level.Info(t.logger).Log("msg", "read lines routine finished", "path", t.path)
+		close(t.done)
+	}()
+	entries := t.handler.Chan()
+
+	f, err := os.Open(t.path)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "error reading file", "path", t.path, "error", err)
+		return
+	}
+	defer f.Close()
+
+	r, err := mountReader(f, t.logger)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "error mounting new reader", "err", err)
+		return
+	}
+
+	level.Info(t.logger).Log("msg", "successfully decompressed file", "path", t.path)
+
+	scanner := bufio.NewScanner(r)
+	for i := 0; ; i++ {
+		if !scanner.Scan() {
+			break
+		}
+
+		if scanner.Err() != nil {
+			break
+		}
+
+		if i <= int(t.position) {
+			// skip already seen lines.
+			continue
+		}
+
+		text := scanner.Text()
+		var finalText string
+		if t.decoder != nil {
+			var err error
+			finalText, err = t.convertToUTF8(text)
+			if err != nil {
+				level.Debug(t.logger).Log("msg", "failed to convert encoding", "error", err)
+				t.metrics.encodingFailures.WithLabelValues(t.path).Inc()
+				finalText = fmt.Sprintf("the requested encoding conversion for this line failed in Promtail/Grafana Agent: %s", err.Error())
+			}
+		} else {
+			finalText = text
+		}
+
+		t.metrics.readLines.WithLabelValues(t.path).Inc()
+
+		entries <- api.Entry{
+			Labels: model.LabelSet{},
+			Entry: logproto.Entry{
+				Timestamp: time.Now(),
+				Line:      finalText,
+			},
+		}
+
+		t.size = int64(unsafe.Sizeof(finalText))
+		t.position++
+
+		if err != nil {
+			break
+		}
```

Comment:
Where will this `err` be set?
Line `199` seems unnecessary?

Reply:
Oops, I think I had a brain fog :rofl: 
Fixed by https://github.com/grafana/loki/pull/6708/commits/459d1cc04811058b3cdb47c56af1b068c9040308